contracts/verifier.sol

```diff
@@ -0,0 +1,260 @@
+            s.lowStep = step;
+        } else {
+            // this next step must be in the correct range
+            require(step > s.lowStep && step < s.highStep);
+
+            // if this is NOT the first query, update the steps and assign the correct hash
+            // (if this IS the first query, we just want to initialize medStep and medHash)
+            if (!isFirstStep) {
+                if (step < s.medStep) {
+                    // if we're iterating lower,
+                    //   the new highest is the current middle
+                    s.highStep = s.medStep;
+                    s.highHash = s.medHash;
+                } else if (step > s.medStep) {
+                    // if we're iterating upwards,
+                    //   the new lowest is the current middle
+                    s.lowStep = s.medStep;
+                    s.lowHash = s.medHash;
+                } else {
+                    // and if we're requesting the midStep that we've already requested,
+                    //   there's nothing to do.
+                    // @TODO(shrugs) - should this revert?
+                }
+            }
+
+            s.medStep = step;
+            s.medHash = bytes32(0);
+        }
+        s.lastChallengerMessage = now;
+        NewQuery(sessionId);
+    }
+
+    function respond(uint sessionId, uint step, bytes32 hash)
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        // Require step to avoid replay problems
+        require(step == s.medStep);
+        if (hash == 0) {
+            // Special "fold" signal
+            // @TODO(shrugs) - is this correct? should the claimant not fold here?
+            challengerConvicted(sessionId);
+            return;
+        }
+
+        // record the claimed hash
+        s.medHash = hash;
+        s.lastClaimantMessage = now;
+
+        // notify watchers
+        NewResponse(sessionId);
+    }
+
+    function performStepVerification(
+        uint sessionId,
+        bytes preValue,
+        bytes postValue,
+        bytes proofs,
+        address claimManager
+    )
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        require(s.lowStep + 1 == s.highStep);
+        // ^ must be at the end of the binary search according to the smart contract
+
+        if (keccak256(preValue) != s.lowHash) { claimantConvicted(sessionId); return; }
+        if (keccak256(postValue) != s.highHash) { return claimantConvicted(sessionId); return; }
+        // ^ if the claimant supplied anything other than what we know, convict them
+        // @TODO(shrugs) - why do we even have this if we already know this information?
+
+        ClaimManager cm = ClaimManager(claimManager);
+        if (performStepVerificationSpecific(s, s.lowStep, preValue, postValue, proofs)) {
+            cm.claimDecided(sessionId, s.claimant, s.challenger);
```

Comment:
This should be moved into the `convicted` functions.

Reply:
because otherwise, the timout does not call it.

Reply:
Added an issue: https://github.com/TrueBitFoundation/scrypt-interactive/issues/41